File: bubbler_app/lib/data/models/preferences.dart
import 'package:bubbler_app/data/models/topics.dart';

enum PreferenceType {
  preferred,
  blacklisted;

  static PreferenceType fromJson(String value) {
    return PreferenceType.values.firstWhere(
      (e) => e.name == value,
      orElse: () => throw FormatException('Unknown preference type: $value'),
    );
  }

  String toJson() => name;
}

class TopicPreference {
  const TopicPreference({
    required this.topic,
    required this.preferenceType,
  });

  factory TopicPreference.fromJson(Map<String, dynamic> json) {
    return TopicPreference(
      topic: json['topic'] as String,
      preferenceType:
          PreferenceType.fromJson(json['preference_type'] as String),
    );
  }

  final String topic;
  final PreferenceType preferenceType;

  Map<String, dynamic> toJson() {
    return {
      'topic': topic,
      'preference_type': preferenceType.toJson(),
    };
  }

  @override
  bool operator ==(Object other) {
    return other is TopicPreference &&
        other.topic == topic &&
        other.preferenceType == preferenceType;
  }

  @override
  int get hashCode => Object.hash(topic, preferenceType);
}

/// Feed composition weights (`similar` / `graph` / `opposite` / `random`).
class FeedStrategyWeights {
  const FeedStrategyWeights({
    required this.similar,
    required this.graph,
    required this.opposite,
    required this.random,
  });

  factory FeedStrategyWeights.fromJson(Map<String, dynamic> json) {
    double read(String key, double fallback) {
      final value = json[key];
      if (value is num) return value.toDouble();
      return fallback;
    }

    return FeedStrategyWeights(
      similar: read('similar', FeedStrategyWeights.defaults.similar),
      graph: read('graph', FeedStrategyWeights.defaults.graph),
      opposite: read('opposite', FeedStrategyWeights.defaults.opposite),
      random: read('random', FeedStrategyWeights.defaults.random),
    );
  }

  static const FeedStrategyWeights defaults = FeedStrategyWeights(
    similar: 0.4,
    graph: 0.25,
    opposite: 0.2,
    random: 0.15,
  );

  final double similar;
  final double graph;
  final double opposite;
  final double random;

  double get total => similar + graph + opposite + random;

  FeedStrategyWeights normalized() {
    final clamped = FeedStrategyWeights(
      similar: similar.clamp(0.0, 1.0),
      graph: graph.clamp(0.0, 1.0),
      opposite: opposite.clamp(0.0, 1.0),
      random: random.clamp(0.0, 1.0),
    );

    final sum = clamped.total;
    if (sum <= 0) return FeedStrategyWeights.defaults;

    return FeedStrategyWeights(
      similar: clamped.similar / sum,
      graph: clamped.graph / sum,
      opposite: clamped.opposite / sum,
      random: clamped.random / sum,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'similar': similar,
      'graph': graph,
      'opposite': opposite,
      'random': random,
    };
  }

  @override
  bool operator ==(Object other) {
    return other is FeedStrategyWeights &&
        other.similar == similar &&
        other.graph == graph &&
        other.opposite == opposite &&
        other.random == random;
  }

  @override
  int get hashCode => Object.hash(similar, graph, opposite, random);
}

/// PUT body for preference updates (no `user_id`).
class PreferencesUpdatePayload {
  const PreferencesUpdatePayload({
    required this.diversityTolerance,
    required this.randomness,
    required this.topicPreferences,
    required this.useViewTime,
    required this.viewTimeWeight,
    required this.useRecency,
    required this.aiTopicDetection,
    required this.strategyWeights,
  });

  factory PreferencesUpdatePayload.fromJson(Map<String, dynamic> json) {
    return PreferencesUpdatePayload(
      diversityTolerance: (json['diversity_tolerance'] as num).toDouble(),
      randomness: (json['randomness'] as num).toDouble(),
      topicPreferences: (json['topic_preferences'] as List<dynamic>? ?? const [])
          .map((e) => TopicPreference.fromJson(e as Map<String, dynamic>))
          .toList(),
      useViewTime: json['use_view_time'] as bool? ?? false,
      viewTimeWeight: (json['view_time_weight'] as num?)?.toDouble() ?? 0.1,
      useRecency: json['use_recency'] as bool? ?? true,
      aiTopicDetection: json['ai_topic_detection'] as bool? ?? false,
      strategyWeights: FeedStrategyWeights.fromJson(
        Map<String, dynamic>.from(json['strategy_weights'] as Map? ?? const {}),
      ),
    );
  }

  final double diversityTolerance;
  final double randomness;
  final List<TopicPreference> topicPreferences;
  final bool useViewTime;
  final double viewTimeWeight;
  final bool useRecency;
  final bool aiTopicDetection;
  final FeedStrategyWeights strategyWeights;

  Map<String, dynamic> toJson() {
    return {
      'diversity_tolerance': diversityTolerance,
      'randomness': randomness,
      'topic_preferences': topicPreferences.map((t) => t.toJson()).toList(),
      'use_view_time': useViewTime,
      'view_time_weight': viewTimeWeight,
      'use_recency': useRecency,
      'ai_topic_detection': aiTopicDetection,
      'strategy_weights': strategyWeights.toJson(),
    };
  }
}

/// User recommendation preferences from `GET /user/me/preferences`.
class UserPreferences {
  const UserPreferences({
    required this.userId,
    required this.diversityTolerance,
    required this.randomness,
    required this.topicPreferences,
    required this.useViewTime,
    required this.viewTimeWeight,
    required this.useRecency,
    required this.aiTopicDetection,
    required this.strategyWeights,
  });

  factory UserPreferences.fromJson(Map<String, dynamic> json) {
    return UserPreferences(
      userId: json['user_id'] as int,
      diversityTolerance: (json['diversity_tolerance'] as num).toDouble(),
      randomness: (json['randomness'] as num).toDouble(),
      topicPreferences: (json['topic_preferences'] as List<dynamic>? ?? const [])
          .map((e) => TopicPreference.fromJson(e as Map<String, dynamic>))
          .toList(),
      useViewTime: json['use_view_time'] as bool? ?? false,
      viewTimeWeight: (json['view_time_weight'] as num?)?.toDouble() ?? 0.1,
      useRecency: json['use_recency'] as bool? ?? true,
      aiTopicDetection: json['ai_topic_detection'] as bool? ?? false,
      strategyWeights: FeedStrategyWeights.fromJson(
        Map<String, dynamic>.from(json['strategy_weights'] as Map? ?? const {}),
      ),
    );
  }

  static const UserPreferences placeholder = UserPreferences(
    userId: 0,
    diversityTolerance: 0.4,
    randomness: 0.4,
    topicPreferences: [],
    useViewTime: false,
    viewTimeWeight: 0.1,
    useRecency: true,
    aiTopicDetection: false,
    strategyWeights: FeedStrategyWeights.defaults,
  );

  /// Built-in algorithm defaults (matches backend `default_user_prefs`).
  static UserPreferences systemDefaults({required int userId}) {
    return UserPreferences(
      userId: userId,
      diversityTolerance: 0.4,
      randomness: 0.4,
      topicPreferences: const [],
      useViewTime: false,
      viewTimeWeight: 0.1,
      useRecency: true,
      aiTopicDetection: false,
      strategyWeights: FeedStrategyWeights.defaults,
    );
  }

  final int userId;
  final double diversityTolerance;
  final double randomness;
  final List<TopicPreference> topicPreferences;
  final bool useViewTime;
  final double viewTimeWeight;
  final bool useRecency;
  final bool aiTopicDetection;
  final FeedStrategyWeights strategyWeights;

  List<String> get preferredTopics => topicPreferences
      .where((t) => t.preferenceType == PreferenceType.preferred)
      .map((t) => t.topic)
      .toList();

  List<String> get blacklistedTopics => topicPreferences
      .where((t) => t.preferenceType == PreferenceType.blacklisted)
      .map((t) => t.topic)
      .toList();

  PreferencesUpdatePayload get updatePayload => PreferencesUpdatePayload(
        diversityTolerance: diversityTolerance,
        randomness: randomness,
        topicPreferences: topicPreferences,
        useViewTime: useViewTime,
        viewTimeWeight: viewTimeWeight,
        useRecency: useRecency,
        aiTopicDetection: aiTopicDetection,
        strategyWeights: strategyWeights,
      );

  UserPreferences copyWith({
    int? userId,
    double? diversityTolerance,
    double? randomness,
    List<TopicPreference>? topicPreferences,
    bool? useViewTime,
    double? viewTimeWeight,
    bool? useRecency,
    bool? aiTopicDetection,
    FeedStrategyWeights? strategyWeights,
  }) {
    return UserPreferences(
      userId: userId ?? this.userId,
      diversityTolerance: diversityTolerance ?? this.diversityTolerance,
      randomness: randomness ?? this.randomness,
      topicPreferences: topicPreferences ?? this.topicPreferences,
      useViewTime: useViewTime ?? this.useViewTime,
      viewTimeWeight: viewTimeWeight ?? this.viewTimeWeight,
      useRecency: useRecency ?? this.useRecency,
      aiTopicDetection: aiTopicDetection ?? this.aiTopicDetection,
      strategyWeights: strategyWeights ?? this.strategyWeights,
    );
  }

  UserPreferences updatePreferredTopics(List<String> topics) {
    final preferred = TopicPreferenceList.cleaned(topics)
        .map((t) => TopicPreference(
              topic: t,
              preferenceType: PreferenceType.preferred,
            ))
        .toList();
    final preferredKeys = preferred.map((t) => t.topic.toLowerCase()).toSet();
    // Preferring a topic must clear any blacklist entry in the same update.
    final blacklisted = topicPreferences
        .where(
          (t) =>
              t.preferenceType == PreferenceType.blacklisted &&
              !preferredKeys.contains(t.topic.toLowerCase()),
        )
        .toList();
    return copyWith(
      topicPreferences: _mergeTopicPreferences(
        preferred: preferred,
        blacklisted: blacklisted,
      ),
    );
  }

  UserPreferences updateBlacklistedTopics(List<String> topics) {
    final blacklisted = TopicPreferenceList.cleaned(topics)
        .map((t) => TopicPreference(
              topic: t,
              preferenceType: PreferenceType.blacklisted,
            ))
        .toList();
    final blacklistedKeys =
        blacklisted.map((t) => t.topic.toLowerCase()).toSet();
    // Blacklisting a topic must clear any preferred entry in the same update.
    final preferred = topicPreferences
        .where(
          (t) =>
              t.preferenceType == PreferenceType.preferred &&
              !blacklistedKeys.contains(t.topic.toLowerCase()),
        )
        .toList();
    return copyWith(
      topicPreferences: _mergeTopicPreferences(
        preferred: preferred,
        blacklisted: blacklisted,
      ),
    );
  }

  UserPreferences preferTopic(String topic) {
    return updatePreferredTopics(
      TopicPreferenceList.add(topic, preferredTopics),
    ).updateBlacklistedTopics(
      TopicPreferenceList.remove(topic, blacklistedTopics),
    );
  }

  UserPreferences unpreferTopic(String topic) {
    return updatePreferredTopics(
      TopicPreferenceList.remove(topic, preferredTopics),
    );
  }

  UserPreferences blacklistTopic(String topic) {
    return updatePreferredTopics(
      TopicPreferenceList.remove(topic, preferredTopics),
    ).updateBlacklistedTopics(
      TopicPreferenceList.add(topic, blacklistedTopics),
    );
  }

  UserPreferences unblacklistTopic(String topic) {
    return updateBlacklistedTopics(
      TopicPreferenceList.remove(topic, blacklistedTopics),
    );
  }

  UserPreferences sanitized() {
    final preferred = TopicPreferenceList.cleaned(preferredTopics);
    final blacklist = TopicPreferenceList.cleaned(blacklistedTopics)
        .where(
          (blacklistedTopic) => !preferred.any(
            (p) => p.toLowerCase() == blacklistedTopic.toLowerCase(),
          ),
        )
        .toList();

    return UserPreferences(
      userId: userId,
      diversityTolerance: diversityTolerance.clamp(0.0, 1.0),
      randomness: randomness.clamp(0.0, 1.0),
      topicPreferences: _mergeTopicPreferences(
        preferred: preferred
            .map(
              (t) => TopicPreference(
                topic: t,
                preferenceType: PreferenceType.preferred,
              ),
            )
            .toList(),
        blacklisted: blacklist
            .map(
              (t) => TopicPreference(
                topic: t,
                preferenceType: PreferenceType.blacklisted,
              ),
            )
            .toList(),
      ),
      useViewTime: useViewTime,
      viewTimeWeight: viewTimeWeight.clamp(0.0, 1.0),
      useRecency: useRecency,
      aiTopicDetection: aiTopicDetection,
      strategyWeights: strategyWeights.normalized(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'diversity_tolerance': diversityTolerance,
      'randomness': randomness,
      'topic_preferences': topicPreferences.map((t) => t.toJson()).toList(),
      'use_view_time': useViewTime,
      'view_time_weight': viewTimeWeight,
      'use_recency': useRecency,
      'ai_topic_detection': aiTopicDetection,
      'strategy_weights': strategyWeights.toJson(),
    };
  }

  static List<TopicPreference> _mergeTopicPreferences({
    required List<TopicPreference> preferred,
    required List<TopicPreference> blacklisted,
  }) {
    final seen = <String>{};
    final merged = <TopicPreference>[];

    for (final pref in [...preferred, ...blacklisted]) {
      final key = pref.topic.toLowerCase();
      if (!seen.add(key)) continue;
      merged.add(pref);
    }

    merged.sort(
      (a, b) => a.topic.toLowerCase().compareTo(b.topic.toLowerCase()),
    );
    return merged;
  }

  @override
  bool operator ==(Object other) {
    if (other is! UserPreferences) return false;
    if (other.userId != userId ||
        other.diversityTolerance != diversityTolerance ||
        other.randomness != randomness ||
        other.useViewTime != useViewTime ||
        other.viewTimeWeight != viewTimeWeight ||
        other.useRecency != useRecency ||
        other.aiTopicDetection != aiTopicDetection ||
        other.strategyWeights != strategyWeights) {
      return false;
    }
    if (other.topicPreferences.length != topicPreferences.length) return false;
    for (var i = 0; i < topicPreferences.length; i++) {
      if (other.topicPreferences[i] != topicPreferences[i]) return false;
    }
    return true;
  }

  @override
  int get hashCode => Object.hash(
        userId,
        diversityTolerance,
        randomness,
        Object.hashAll(topicPreferences),
        useViewTime,
        viewTimeWeight,
        useRecency,
        aiTopicDetection,
        strategyWeights,
      );
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/model/Preferences.kt
package com.bubbler.android.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

@Serializable
enum class FeedPreset {
    @SerialName("stay_in_lane")
    STAY_IN_LANE,

    @SerialName("cross_pollinate")
    CROSS_POLLINATE,

    @SerialName("wild_walk")
    WILD_WALK,

    @SerialName("custom")
    CUSTOM,
}

enum class CompositionTier {
    TOPIC,
    POST,
}

@Serializable
enum class PreferenceType {
    @SerialName("preferred")
    PREFERRED,

    @SerialName("blacklisted")
    BLACKLISTED,
}

@Serializable
data class TopicPreference(
    val topic: String,
    @SerialName("preference_type")
    val preferenceType: PreferenceType,
)

@Serializable
data class CompositionWeights(
    val similar: Double = DEFAULT.similar,
    val opposite: Double = DEFAULT.opposite,
    val surprise: Double = DEFAULT.surprise,
) {
    val total: Double get() = similar + opposite + surprise

    fun normalized(): CompositionWeights {
        val clamped = CompositionWeights(
            similar = similar.clamped(0.0, 1.0),
            opposite = opposite.clamped(0.0, 1.0),
            surprise = surprise.clamped(0.0, 1.0),
        )
        val sum = clamped.total
        if (sum <= 0.0) return DEFAULT
        return CompositionWeights(
            similar = clamped.similar / sum,
            opposite = clamped.opposite / sum,
            surprise = clamped.surprise / sum,
        )
    }

    companion object {
        val DEFAULT = CompositionWeights(
            similar = 0.55,
            opposite = 0.15,
            surprise = 0.30,
        )

        private const val MATCH_EPSILON = 0.02

        fun presetWeights(preset: FeedPreset, tier: CompositionTier): CompositionWeights =
            when (preset) {
                FeedPreset.STAY_IN_LANE -> DEFAULT
                FeedPreset.CROSS_POLLINATE -> if (tier == CompositionTier.TOPIC) {
                    CompositionWeights(similar = 0.15, opposite = 0.55, surprise = 0.30)
                } else {
                    CompositionWeights(similar = 0.55, opposite = 0.15, surprise = 0.30)
                }
                FeedPreset.WILD_WALK -> CompositionWeights(
                    similar = 0.15,
                    opposite = 0.25,
                    surprise = 0.60,
                )
                FeedPreset.CUSTOM -> DEFAULT
            }

        fun matches(left: CompositionWeights, right: CompositionWeights): Boolean {
            val a = left.normalized()
            val b = right.normalized()
            return abs(a.similar - b.similar) <= MATCH_EPSILON &&
                abs(a.opposite - b.opposite) <= MATCH_EPSILON &&
                abs(a.surprise - b.surprise) <= MATCH_EPSILON
        }
    }
}

@Serializable
data class UserPreferences(
    @SerialName("user_id")
    val userId: Int,
    @SerialName("feed_preset")
    val feedPreset: FeedPreset = FeedPreset.STAY_IN_LANE,
    @SerialName("topic_composition")
    val topicComposition: CompositionWeights = CompositionWeights.presetWeights(
        FeedPreset.STAY_IN_LANE,
        CompositionTier.TOPIC,
    ),
    @SerialName("post_composition")
    val postComposition: CompositionWeights = CompositionWeights.presetWeights(
        FeedPreset.STAY_IN_LANE,
        CompositionTier.POST,
    ),
    @SerialName("topic_preferences")
    val topicPreferences: List<TopicPreference> = emptyList(),
    @SerialName("use_view_time")
    val useViewTime: Boolean = false,
    @SerialName("view_time_weight")
    val viewTimeWeight: Double = 0.1,
    @SerialName("use_recency")
    val useRecency: Boolean = false,
    @SerialName("ai_topic_detection")
    val aiTopicDetection: Boolean = false,
) {
    val preferredTopics: List<String>
        get() = topicPreferences
            .filter { it.preferenceType == PreferenceType.PREFERRED }
            .map { it.topic }

    val blacklistedTopics: List<String>
        get() = topicPreferences
            .filter { it.preferenceType == PreferenceType.BLACKLISTED }
            .map { it.topic }

    val updatePayload: PreferencesUpdatePayload
        get() = PreferencesUpdatePayload(
            feedPreset = feedPreset,
            topicComposition = topicComposition,
            postComposition = postComposition,
            topicPreferences = topicPreferences,
            useViewTime = useViewTime,
            viewTimeWeight = viewTimeWeight,
            useRecency = useRecency,
            aiTopicDetection = aiTopicDetection,
        )

    fun applyPreset(preset: FeedPreset): UserPreferences {
        if (preset == FeedPreset.CUSTOM) {
            return copy(feedPreset = FeedPreset.CUSTOM)
        }
        return copy(
            feedPreset = preset,
            topicComposition = CompositionWeights.presetWeights(preset, CompositionTier.TOPIC),
            postComposition = CompositionWeights.presetWeights(preset, CompositionTier.POST),
        )
    }

    fun detectPreset(): UserPreferences {
        for (preset in SELECTABLE_PRESETS) {
            val topic = CompositionWeights.presetWeights(preset, CompositionTier.TOPIC)
            val post = CompositionWeights.presetWeights(preset, CompositionTier.POST)
            if (CompositionWeights.matches(topicComposition, topic) &&
                CompositionWeights.matches(postComposition, post)
            ) {
                return copy(feedPreset = preset)
            }
        }
        return copy(feedPreset = FeedPreset.CUSTOM)
    }

    fun updatePreferredTopics(topics: List<String>): UserPreferences {
        val preferred = TopicPreferenceList.cleaned(topics).map {
            TopicPreference(it, PreferenceType.PREFERRED)
        }
        val preferredKeys = preferred.map { it.topic.lowercase() }.toSet()
        val blacklisted = topicPreferences.filter {
            it.preferenceType == PreferenceType.BLACKLISTED &&
                it.topic.lowercase() !in preferredKeys
        }
        return copy(topicPreferences = mergeTopicPreferences(preferred, blacklisted))
    }

    fun updateBlacklistedTopics(topics: List<String>): UserPreferences {
        val blacklisted = TopicPreferenceList.cleaned(topics).map {
            TopicPreference(it, PreferenceType.BLACKLISTED)
        }
        val blacklistedKeys = blacklisted.map { it.topic.lowercase() }.toSet()
        val preferred = topicPreferences.filter {
            it.preferenceType == PreferenceType.PREFERRED &&
                it.topic.lowercase() !in blacklistedKeys
        }
        return copy(topicPreferences = mergeTopicPreferences(preferred, blacklisted))
    }

    fun preferTopic(topic: String): UserPreferences {
        val withPreferred = updatePreferredTopics(
            TopicPreferenceList.add(topic, to = preferredTopics),
        )
        return withPreferred.updateBlacklistedTopics(
            TopicPreferenceList.remove(topic, from = withPreferred.blacklistedTopics),
        )
    }

    fun unpreferTopic(topic: String): UserPreferences =
        updatePreferredTopics(TopicPreferenceList.remove(topic, from = preferredTopics))

    fun blacklistTopic(topic: String): UserPreferences {
        val withoutPreferred = updatePreferredTopics(
            TopicPreferenceList.remove(topic, from = preferredTopics),
        )
        return withoutPreferred.updateBlacklistedTopics(
            TopicPreferenceList.add(topic, to = withoutPreferred.blacklistedTopics),
        )
    }

    fun unblacklistTopic(topic: String): UserPreferences =
        updateBlacklistedTopics(TopicPreferenceList.remove(topic, from = blacklistedTopics))

    fun sanitized(): UserPreferences {
        val preferred = TopicPreferenceList.cleaned(preferredTopics)
        val blacklist = TopicPreferenceList.cleaned(blacklistedTopics)
            .filter { blacklistedTopic ->
                preferred.none { it.equals(blacklistedTopic, ignoreCase = true) }
            }

        var result = UserPreferences(
            userId = userId,
            feedPreset = feedPreset,
            topicComposition = topicComposition.normalized(),
            postComposition = postComposition.normalized(),
            topicPreferences = mergeTopicPreferences(
                preferred = preferred.map { TopicPreference(it, PreferenceType.PREFERRED) },
                blacklisted = blacklist.map { TopicPreference(it, PreferenceType.BLACKLISTED) },
            ),
            useViewTime = useViewTime,
            viewTimeWeight = viewTimeWeight.clamped(0.0, 1.0),
            useRecency = useRecency,
            aiTopicDetection = aiTopicDetection,
        )
        result = if (result.feedPreset != FeedPreset.CUSTOM) {
            result.applyPreset(result.feedPreset)
        } else {
            result.detectPreset()
        }
        return result
    }

    companion object {
        val PLACEHOLDER: UserPreferences = systemDefaults(userId = 0)

        val SELECTABLE_PRESETS: List<FeedPreset> = listOf(
            FeedPreset.STAY_IN_LANE,
            FeedPreset.CROSS_POLLINATE,
            FeedPreset.WILD_WALK,
        )

        fun systemDefaults(userId: Int): UserPreferences = UserPreferences(
            userId = userId,
            feedPreset = FeedPreset.STAY_IN_LANE,
            topicComposition = CompositionWeights.presetWeights(
                FeedPreset.STAY_IN_LANE,
                CompositionTier.TOPIC,
            ),
            postComposition = CompositionWeights.presetWeights(
                FeedPreset.STAY_IN_LANE,
                CompositionTier.POST,
            ),
            topicPreferences = emptyList(),
            useViewTime = false,
            viewTimeWeight = 0.1,
            useRecency = false,
            aiTopicDetection = false,
        )

        private fun mergeTopicPreferences(
            preferred: List<TopicPreference>,
            blacklisted: List<TopicPreference>,
        ): List<TopicPreference> {
            val seen = mutableSetOf<String>()
            val merged = mutableListOf<TopicPreference>()
            for (pref in preferred + blacklisted) {
                val key = pref.topic.lowercase()
                if (seen.add(key)) {
                    merged.add(pref)
                }
            }
            return merged.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.topic })
        }
    }
}

@Serializable
data class PreferencesUpdatePayload(
    @SerialName("feed_preset")
    val feedPreset: FeedPreset = FeedPreset.STAY_IN_LANE,
    @SerialName("topic_composition")
    val topicComposition: CompositionWeights = CompositionWeights.DEFAULT,
    @SerialName("post_composition")
    val postComposition: CompositionWeights = CompositionWeights.DEFAULT,
    @SerialName("topic_preferences")
    val topicPreferences: List<TopicPreference>,
    @SerialName("use_view_time")
    val useViewTime: Boolean = false,
    @SerialName("view_time_weight")
    val viewTimeWeight: Double = 0.1,
    @SerialName("use_recency")
    val useRecency: Boolean = false,
    @SerialName("ai_topic_detection")
    val aiTopicDetection: Boolean = false,
)

fun FeedPreset.displayTitle(): String = when (this) {
    FeedPreset.STAY_IN_LANE -> "Stay in lane"
    FeedPreset.CROSS_POLLINATE -> "Cross-pollinate"
    FeedPreset.WILD_WALK -> "Wild walk"
    FeedPreset.CUSTOM -> "Custom"
}

fun FeedPreset.displayDescription(): String = when (this) {
    FeedPreset.STAY_IN_LANE ->
        "Keep exploring within familiar topics and closely related posts."
    FeedPreset.CROSS_POLLINATE ->
        "Jump to different topics while surfacing posts that still feel connected to what you're reading."
    FeedPreset.WILD_WALK ->
        "Maximize variety with unexpected topics and posts."
    FeedPreset.CUSTOM ->
        "Your manually tuned topic and post mix (set in Advanced)."
}

private fun Double.clamped(min: Double, max: Double): Double = min(max(this, min), max)
